From c695a93b339ea41da97859c693dbe7ebb6f9e911 Mon Sep 17 00:00:00 2001
From: Yuvarani V <quic_yuvarani@quicinc.com>
Date: Mon, 4 Dec 2023 00:11:47 +0530
Subject: [PATCH] hostapd: RNR signalling 320MHz BW reduction for 6GHz radio

RnR signaling will not use 320MHz Operating class as
WiFi6/6E STAs will not support decoding 320MHz Opclass. We have
reduced it to 160MHz Operating class

Signed-off-by: Karthik M <quic_karm@quicinc.com>
Signed-off-by: Yuvarani V <quic_yuvarani@quicinc.com>
---
 src/ap/ieee802_11.c | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -8295,15 +8295,20 @@ static u8 * hostapd_eid_rnr_iface(struct
 	if (!(iface->drv_flags & WPA_DRIVER_FLAGS_AP_CSA) || !iface->freq)
 		return eid;
 
-	/* op_class for 5GHz 320MHz bw is not defined in spec. So use op_class of 160MHz */
-	if (is_5ghz_freq(hapd->iface->freq) && ch_width == CONF_OPER_CHWIDTH_320MHZ)
-		op_class = 129;
-	else
-		if (ieee80211_freq_to_channel_ext(iface->freq,
-						  hapd->iconf->secondary_channel,
-						  ch_width, &op_class, &channel) ==
-		    NUM_HOSTAPD_MODES)
-			return eid;
+	if (ch_width == CONF_OPER_CHWIDTH_320MHZ) {
+		/* For 5GHz, 320MHz BW is not defined in spec.
+		 * So use op_class of 160MHz
+		 * As per Draft P802.11Revme_D4.0, WIFI6/6E STA will
+		 * not support 320MHz op_class. So use op_class of 160MHz
+		 * in these cases.
+		 */
+		op_class = is_5ghz_freq(hapd->iface->freq) ? 129 : 134;
+	} else if (ieee80211_freq_to_channel_ext(iface->freq,
+						 hapd->iconf->secondary_channel,
+						 ch_width, &op_class,
+						 &channel) == NUM_HOSTAPD_MODES) {
+		return eid;
+	}
 
 repeat_rnr:
 	start = 0;
